### PR TITLE
Added a secondary Cross (Blue) default at light level 1

### PR DESCRIPTION
**Pull request by ModProg**
I found this useful mainly for strip mining,  the blue cross is the correct position for a new torch in that case. 

ModMenu:
grafik

This commit also includes a fix for Pre-2, if this collides with anything I can remove that.
It's only one change (`getMaximum` to `getMax`)

I developed the feature for 1.15 as well, but I didn't want to create two identical pull requests.

This is mainly something I did for private use, but I cleaned up my code, so in case there is interest in such a feature in LightOverlay I made this pull request

### CODE REVIEW

File: src/main/resources/assets/lightoverlay/lang/en_us.json [hunk 1]
```diff
@@ -10,11 +10,13 @@
   "config.lightoverlay.general": "General",
   "config.lightoverlay.reach": "Reach:",
   "config.lightoverlay.crossLevel": "Light Level:",
+  "config.lightoverlay.crossLevel2": "Light Level Secondary:",
```
Comment (shedaniel) [inline src/main/resources/assets/lightoverlay/lang/en_us.json]:
```suggestion
  "config.lightoverlay.optimalPlacementLevel": "Optimal Placement Level:",
```

File: src/main/resources/assets/lightoverlay/lang/en_us.json [hunk 2]
```diff
@@ -10,11 +10,13 @@
...
   "config.lightoverlay.showNumber": "Show Number Instead:",
   "config.lightoverlay.smoothLines": "Smooth Lines:",
   "config.lightoverlay.underwater": "Display Underwater:",
   "config.lightoverlay.lineWidth": "Line Width:",
   "config.lightoverlay.yellowColor": "Yellow Color:",
   "config.lightoverlay.redColor": "Red Color:",
+  "config.lightoverlay.secondaryColor": "Secondary Color:",
```
Comment (shedaniel) [inline src/main/resources/assets/lightoverlay/lang/en_us.json]:
```suggestion
  "config.lightoverlay.optimalPlacementColor": "Optimal Placement Color:",
```

File: src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java [hunk 3]
```diff
@@ -28,12 +28,14 @@ public Screen getConfigScreenByCloth(Screen parent) {
         ConfigCategory general = builder.getOrCreateCategory(new TranslatableText("config.lightoverlay.general"));
         general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.reach"), LightOverlay.reach, 1, 64).setDefaultValue(12).setTextGetter(integer -> new LiteralText("Reach: " + integer + " Blocks")).setSaveConsumer(integer -> LightOverlay.reach = integer).build());
         general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.crossLevel"), LightOverlay.crossLevel, 0, 15).setDefaultValue(7).setTextGetter(integer -> new LiteralText("Cross Level: " + integer)).setSaveConsumer(integer -> LightOverlay.crossLevel = integer).build());
+        general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.crossLevel2"), LightOverlay.crossLevel2, 0, 15).setDefaultValue(1).setTextGetter(integer -> new LiteralText("Secondary Level: " + integer)).setSaveConsumer(integer -> LightOverlay.crossLevel2 = integer).build());
```
Comment (shedaniel) [inline src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java]:
```suggestion
        general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.optimalPlacementLevel"), LightOverlay.optimalPlacementLevel, 0, 15).setDefaultValue(1).setTextGetter(integer -> new LiteralText("Level: " + integer)).setSaveConsumer(integer -> LightOverlay.optimalPlacementLevel = integer).build());
```

File: src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java [hunk 4]
```diff
@@ -28,12 +28,14 @@ public Screen getConfigScreenByCloth(Screen parent) {
         ConfigCategory general = builder.getOrCreateCategory(new TranslatableText("config.lightoverlay.general"));
         general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.reach"), LightOverlay.reach, 1, 64).setDefaultValue(12).setTextGetter(integer -> new LiteralText("Reach: " + integer + " Blocks")).setSaveConsumer(integer -> LightOverlay.reach = integer).build());
         general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.crossLevel"), LightOverlay.crossLevel, 0, 15).setDefaultValue(7).setTextGetter(integer -> new LiteralText("Cross Level: " + integer)).setSaveConsumer(integer -> LightOverlay.crossLevel = integer).build());
```
Comment (shedaniel) [inline src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java]:
```suggestion
        general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.crossLevel"), LightOverlay.crossLevel, 0, 15).setDefaultValue(7).setTextGetter(integer -> new LiteralText("Level: " + integer)).setSaveConsumer(integer -> LightOverlay.crossLevel = integer).build());
```

File: src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java [hunk 5]
```diff
@@ -28,12 +28,14 @@ public Screen getConfigScreenByCloth(Screen parent) {
...
         general.addEntry(eb.startIntSlider(new TranslatableText("config.lightoverlay.lineWidth"), MathHelper.floor(LightOverlay.lineWidth * 100), 100, 700).setDefaultValue(100).setTextGetter(integer -> new LiteralText("Light Width: " + LightOverlay.FORMAT.format(integer / 100d))).setSaveConsumer(integer -> LightOverlay.lineWidth = integer / 100f).build());
         general.addEntry(eb.startColorField(new TranslatableText("config.lightoverlay.yellowColor"), LightOverlay.yellowColor).setDefaultValue(0xFFFF00).setSaveConsumer(color -> LightOverlay.yellowColor = color).build());
         general.addEntry(eb.startColorField(new TranslatableText("config.lightoverlay.redColor"), LightOverlay.redColor).setDefaultValue(0xFF0000).setSaveConsumer(color -> LightOverlay.redColor = color).build());
+        general.addEntry(eb.startColorField(new TranslatableText("config.lightoverlay.secondaryColor"), LightOverlay.secondaryColor).setDefaultValue(0x0000FF).setSaveConsumer(color -> LightOverlay.secondaryColor = color).build());
```
Comment (shedaniel) [inline src/main/java/me/shedaniel/lightoverlay/fabric/LOModMenuEntry.java]:
```suggestion
        general.addEntry(eb.startColorField(new TranslatableText("config.lightoverlay.optimalPlacementColor"), LightOverlay.optimalPlacementColor).setDefaultValue(0x0000FF).setSaveConsumer(color -> LightOverlay.optimalPlacementColor = color).build());
```

File: src/main/java/me/shedaniel/lightoverlay/fabric/LightOverlay.java [hunk 6]
```diff
@@ -61,11 +61,12 @@
     private static final Identifier DECREASE_LINE_WIDTH_KEYBIND = new Identifier("lightoverlay", "decrease_line_width");
     static int reach = 12;
     static int crossLevel = 7;
+    static int crossLevel2 = 1;
     static boolean showNumber = false;
     static boolean smoothLines = true;
     static boolean underwater = false;
     static float lineWidth = 1.0F;
-    static int yellowColor = 0xFFFF00, redColor = 0xFF0000;
+    static int yellowColor = 0xFFFF00, redColor = 0xFF0000, secondaryColor = 0x0000FF;
```
Comment (shedaniel) [inline src/main/java/me/shedaniel/lightoverlay/fabric/LightOverlay.java]:
```suggestion
    static int yellowColor = 0xFFFF00, redColor = 0xFF0000, optimalPlacementColor = 0x0000FF;
```

**Review by shedaniel — CHANGES_REQUESTED** [review]
Would want to have a boolean toggle to this optimal placement cross (what you call as secondary cross)

Otherwise, change the names and lgtm

**Comment by ModProg** [comment]
It can now be disabled by setting it to "-1".
I didn't name it Optimal placement Level, because of that being 2 for diagonals.
But as the light level can be changed this is something the user can change to his liking.

Should there be indication that the "Placement Level" is a "Light Level" as well?
grafik
grafik

**Comment by shedaniel** [comment]
maybe do `Hint Light Level`, also this should be disabled by default.

Also what is the optimal location?

**Comment by ModProg** [comment]
It is light level 2 diagonally and 1 in a straight line.

grafik

**Comment by ModProg** [comment]
Should I add an Info like that? 
To the slider maybe:
Level: 1 (opt. straight)

**Comment by shedaniel** [comment]
maybe we could overhaul this whole feature to make it not user customizable for the level, but can display the optimal place to place torches and stuff?

**Comment by ModProg** [comment]
That would probably be quite a complex.
For a plane simply placing diagonally every 7 places in a grid fits.

But for complex shapes there are more than one answer, and the calculation would probably make the whole thing lag.
An example with 1 torch behind a wall, the optimal solution would need to place one of the top torches 1 block to the center, so it is not even symmetrical even tho the setup it self is.
grafik
Here is a [mod](https://www.curseforge.com/minecraft/mc-mods/torch-optimizer) showing optimal torch placement, but it's also just a check for light level = 2

**Comment by ModProg** [comment]
The only thing I could think of that could work, would be to check if a placement of the configured level is close enough to the current location, but even that would require to check for "line of sight" to all viable blocks close enough.

**Comment by shedaniel** [comment]
This is good enough then, I will merge this within a few hours

**Review by shedaniel — APPROVED** [review]

**Comment by ModProg** [comment]
Well an algorithm probably working would be:
```java
if(light_level > 7) return None;
for(int placement_level = 0; placement_level < light_level; placement_level++){
    foreach(block in getBlocksAirAboveBlockLessThen7BlocksAway()){
        if(calculateLightWithTorch(block) > 7) return Red;
    }
}
return Blue
```
But I don't know how much performance this would cost.

Blocks that need to be checked, every glass block and everything in between
grafik
This are about 373 blocks

**Comment by ModProg** [comment]
Should I make the pull request for 1.15 as well?

**Comment by shedaniel** [comment]
I have already done that, just didn't push

**Pull request merged**